### ISSUE TITLE
MongoDB: `Inconsistent column data: Failed to convert 'Decimal128(...)' to 'Decimal'.`

### ISSUE THREAD

**Issue by janpio** [prisma/prisma#12099]
When you have a introspected schema with type `Decimal` (E.g. via https://docs.atlas.mongodb.com/sample-data/sample-airbnb/), you get an error like this when you try to query for the data:

```
Error in Prisma Client request: 

Invalid `prisma.listingsAndReviews.findMany()` invocation:
  Inconsistent column data: Failed to convert 'Decimal128(...)' to 'Decimal'.
```

Schema:
```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["mongoDb"]
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

type ListingsAndReviewsAddress {
  country         String
  country_code    String
  government_area String
  location        ListingsAndReviewsAddressLocation
  market          String
  street          String
  suburb          String
}

type ListingsAndReviewsAddressLocation {
  coordinates       Float[]
  is_location_exact Boolean
  type              String
}

type ListingsAndReviewsAvailability {
  availability_30  Int
  availability_365 Int
  availability_60  Int
  availability_90  Int
}

type ListingsAndReviewsHost {
  host_about                String
  host_has_profile_pic      Boolean
  host_id                   String
  host_identity_verified    Boolean
  host_is_superhost         Boolean
  host_listings_count       Int
  host_location             String
  host_name                 String
  host_neighbourhood        String
  host_picture_url          String
  host_response_rate        Int?
  host_response_time        String?
  host_thumbnail_url        String
  host_total_listings_count Int
  host_url                  String
  host_verifications        String[]
}

type ListingsAndReviewsImages {
  medium_url     String
  picture_url    String
  thumbnail_url  String
  xl_picture_url String
}

type ListingsAndReviewsReviewScores {
  review_scores_accuracy      Int?
  review_scores_checkin       Int?
  review_scores_cleanliness   Int?
  review_scores_communication Int?
  review_scores_location      Int?
  review_scores_rating        Int?
  review_scores_value         Int?
}

type ListingsAndReviewsReviews {
  id            String   @map("_id")
  comments      String?
  date          DateTime @db.Date
  listing_id    String
  reviewer_id   String
  reviewer_name String
}

model listingsAndReviews {
  id                    String                         @id @map("_id")
  access                String
  accommodates          Int
  address               ListingsAndReviewsAddress
  amenities             String[]
  availability          ListingsAndReviewsAvailability
  bathrooms             Decimal?
  bed_type              String
  bedrooms              Int?
  beds                  Int?
  calendar_last_scraped DateTime                       @db.Date
  cancellation_policy   String
  cleaning_fee          Decimal?
  description           String
  extra_people          Decimal
  first_review          DateTime?                      @db.Date
  guests_included       Decimal
  host                  ListingsAndReviewsHost
  house_rules           String
  images                ListingsAndReviewsImages
  interaction           String
  last_review           DateTime?                      @db.Date
  last_scraped          DateTime                       @db.Date
  listing_url           String
  maximum_nights        String
  minimum_nights        String
  monthly_price         Json?
  name                  String
  neighborhood_overview String
  notes                 String
  number_of_reviews     Int
  price                 Decimal
  property_type         String
  review_scores         ListingsAndReviewsReviewScores
  reviews               ListingsAndReviewsReviews[]
  reviews_per_month     Int?
  room_type             String
  security_deposit      Decimal?
  space                 String
  summary               String
  transit               String
  weekly_price          Decimal?

  @@index([property_type, room_type, beds], map: "property_type_1_room_type_1_beds_1")
  @@index([name], map: "name_1")
}
```

(Workaround of defining the field as `Json` also does not work because of https://github.com/prisma/prisma/issues/12097)

**Comment by matthewmueller** [comment]
Deciding whether or not to support this depends on: https://github.com/prisma/prisma-engines/issues/2600

**Comment by matthewmueller** [comment]
On second thought, It seems like introspection should at least work: https://github.com/mongodb/bson-rust/issues/53#issuecomment-894465664

**Comment by janpio** [comment]
Introspection works fine, the `listingsAndReviews` above has many `Decimal` columns - but any interaction with them leads to the reported error.

**Comment by Jolg42** [comment]
`Decimal` is now invalid in our schema for MongoDB

Example schema.prisma validation error
```
Error validating field `bathrooms` in model `listingsAndReviews`: Field `bathrooms` in model `listingsAndReviews` can't be of type Decimal. The current connector does not support the Decimal type.
```

We should add support for it

**Comment by ChristophP** [comment]
This doesn't seem to be specific to Decimal . I'm seeing a similar error for an Int field that has string data.

Schema
```
model test {
  id       String             @id @default(auto()) @map("_id") @db.ObjectId
  name     String
  age      Int
}
```

Value
![image](https://user-images.githubusercontent.com/5880236/196009168-a05a56ae-99fb-480a-a3f3-9074487974ad.png)

Query
```js
  const test = await prisma.test.findMany();
```

Error
```
Inconsistent column data: Failed to convert '"31"' to 'Int' for the field 'age'.: {"code":"P2023","clientVersion":"4.4.0","meta":{"message":"Failed to convert '\"31\"' to 'Int' for the field 'age'."},"stack":"Error: \nInvalid `prisma.test.findMany()` invocation:\n\n\nInconsistent column data: Failed to convert '\"31\"' to 'Int' for the field 'age'.\n    at RequestHandler.handleRequestError (/app/node_modules/@prisma/client/runtime/index.js:30851:13)\n    at RequestHandler.request (/app/node_modules/@prisma/client/runtime/index.js:30834:12)\n    at async PrismaClient._request (/app/node_modules/@prisma/client/runtime/index.js:31812:16)\n    at async file:///app/src/controllers/calendar.controller.js:351:16"}
```

**Comment by janpio** [comment]
Open a new issue for this please @ChristophP - this issue _is_ specific to Decimal and currently blocked by us not supporting Decimal for MongoDB at all.

**Comment by ChristophP** [comment]
sure, will do

**Comment by janpio** [comment]
Closing as I created a proper feature request for this: https://github.com/prisma/prisma/issues/16341
